Security_Engine/Compute/CameraFieldOfView.cs

```diff
@@ -0,0 +1,351 @@
+            }
+            for (int i = 0; i < obstacles.Count; i++)
+            {
+                if (obstacles[i].IsContaining(new List<Point> { line.End }))
+                {
+                    lineObstDict.Add(line, obstacles[i]);
+                    return lineObstDict;
+                }
+                else if (i == obstacles.Count - 1)
+                {
+                    lineObstDict.Add(line, null);
+                }
+            }
+
+            return lineObstDict;
+        }
+
+        /***************************************************/
+
+        [Description("Create points chain for the camera field of view.")]
+        [Input("lineObstacledictionary", "Dictionary of visible lines and obstacles.")]
+        [Input("cameraLocation", "Location of the camera device.")]
+        [Input("radius", "Radius of the camera view cone.")]
+        [Output("pointsChain", "Points chain of the camera field of view.")]
+        private static List<Point> PointsChain(this List<Dictionary<Line, Polyline>> lineObstacledictionary, Point cameraLocation, double radius, double tolerance)
+        {
+            List<Point> pointsChain = new List<Point>();
+            pointsChain.Add(cameraLocation);
+            Point lastPoint = cameraLocation;
+            for (int i = 0; i < lineObstacledictionary.Count; i++)
+            {
+                Line line = lineObstacledictionary[i].Keys.First();
+                if (line.Start.IsEqual(cameraLocation))
+                {
+                    Point point = line.End;
+                    pointsChain.Add(point);
+                    lastPoint = point;
+                }
+                else
+                {
+                    Point pt1 = line.Start;
+                    Point pt2 = line.End;
+
+                    if (lastPoint.Distance(pt1) < lastPoint.Distance(pt2) && !((Math.Abs(lastPoint.Distance(cameraLocation) - radius) < tolerance) && (Math.Abs(pt2.Distance(cameraLocation) - radius) < tolerance)) && i > 0)
+                    {
+                        Polyline obst = lineObstacledictionary[i].Values.First();
+                        Polyline lastObst = lineObstacledictionary[i - 1].Values.First();
+                        if (!lastPoint.IsEqual(cameraLocation) && lastObst == obst)
+                        {
+                            pointsChain.Add(pt2);
+                            pointsChain.Add(pt1);
+                            lastPoint = pt1;
+                            continue;
+                        }
+                        pointsChain.Add(pt1);
+                        pointsChain.Add(pt2);
+                        lastPoint = pt2;
+                    }
+                    else
+                    {
+                        pointsChain.Add(pt2);
+                        pointsChain.Add(pt1);
+                        lastPoint = pt1;
+                    }
+                }
+            }
+            pointsChain = pointsChain.CullDuplicates(tolerance);
```

Comment:
From what I remember, this may change the order of points in the collection - please keep that in mind.